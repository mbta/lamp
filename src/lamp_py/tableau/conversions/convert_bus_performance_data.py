import polars as pl
from pyarrow import Table
import pyarrow


def convert_bus_recent_to_tableau_compatible_schema(
    input_schema: pyarrow.schema, overrides: dict = None, exclude: list = None
) -> pyarrow.schema:
    """
    generic converter for known types (dates, timezones, etc)
    overrides to correct the rest
    overrides_exclude -
    """
    # Create a map to store the modified fields that can't be auto-inferred
    auto_schema = []
    # Loop through the schema fields and apply changes
    for field in input_schema:

        if exclude is not None and field.name in exclude:
            continue
        if overrides is not None and field.name in overrides.keys():
            new_field = field.with_type(overrides[field.name])
            auto_schema.append(new_field)
            continue
        if isinstance(field.type, pyarrow.TimestampType):
            if field.type.tz is not None:
                # strip tz
                new_field = field.with_type(pyarrow.timestamp(unit=field.type.unit))
                auto_schema.append(new_field)
                continue

        # default - no changes, add the field again
        auto_schema.append(field)

    return pyarrow.schema(auto_schema)

def apply_bus_analysis_conversions(polars_df: pl.DataFrame) -> Table:
    """
    Function to apply final conversions to lamp data before outputting for tableau consumption
    """
    # Convert datetime to Eastern Time
    polars_df = polars_df.with_columns(
        pl.col("stop_arrival_dt").dt.convert_time_zone(time_zone="America/New_York").dt.replace_time_zone(None),
        pl.col("stop_departure_dt").dt.convert_time_zone(time_zone="America/New_York").dt.replace_time_zone(None),
        pl.col("gtfs_travel_to_dt").dt.convert_time_zone(time_zone="America/New_York").dt.replace_time_zone(None),
        pl.col("tm_scheduled_time_dt").dt.convert_time_zone(time_zone="America/New_York").dt.replace_time_zone(None),
        pl.col("tm_actual_arrival_dt").dt.convert_time_zone(time_zone="America/New_York").dt.replace_time_zone(None),
        pl.col("tm_actual_departure_dt").dt.convert_time_zone(time_zone="America/New_York").dt.replace_time_zone(None),
        pl.col("gtfs_sort_dt").dt.convert_time_zone(time_zone="America/New_York").dt.replace_time_zone(None),
        pl.col("gtfs_departure_dt").dt.convert_time_zone(time_zone="America/New_York").dt.replace_time_zone(None),
        pl.col("gtfs_arrival_dt").dt.convert_time_zone(time_zone="America/New_York").dt.replace_time_zone(None),
    )

    # Convert seconds columns to be aligned with Eastern Time
    polars_df = polars_df.with_columns(
        (pl.col("gtfs_travel_to_dt") - pl.col("service_date").str.strptime(pl.Date, "%Y%m%d"))
        .dt.total_seconds()
        .alias("gtfs_travel_to_seconds"),
        (pl.col("stop_arrival_dt") - pl.col("service_date").str.strptime(pl.Date, "%Y%m%d"))
        .dt.total_seconds()
        .alias("stop_arrival_seconds"),
        (pl.col("stop_departure_dt") - pl.col("service_date").str.strptime(pl.Date, "%Y%m%d"))
        .dt.total_seconds()
        .alias("stop_departure_seconds"),
    )

    polars_df = polars_df.with_columns(pl.col("service_date").str.strptime(pl.Date, "%Y%m%d", strict=False))

    return polars_df.to_arrow()
