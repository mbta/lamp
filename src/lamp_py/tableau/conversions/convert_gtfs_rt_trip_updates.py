import pyarrow
import polars as pl

from lamp_py.utils.filter_bank import HeavyRailFilter, LightRailFilter

gtfs_rt_trip_updates_processed_schema = pyarrow.schema(
    [
        ("id", pyarrow.large_string()),
        ("trip_update.trip.trip_id", pyarrow.large_string()),
        ("trip_update.trip.route_id", pyarrow.large_string()),
        ("trip_update.trip.direction_id", pyarrow.uint8()),
        ("trip_update.trip.start_time", pyarrow.large_string()),
        ("trip_update.trip.start_date", pyarrow.large_string()),
        ("trip_update.trip.schedule_relationship", pyarrow.large_string()),
        ("trip_update.trip.route_pattern_id", pyarrow.large_string()),
        ("trip_update.trip.tm_trip_id", pyarrow.large_string()),
        ("trip_update.trip.overload_id", pyarrow.int64()),
        ("trip_update.trip.overload_offset", pyarrow.int64()),
        ("trip_update.trip.revenue", pyarrow.bool_()),
        ("trip_update.trip.last_trip", pyarrow.bool_()),
        ("trip_update.vehicle.id", pyarrow.large_string()),
        ("trip_update.vehicle.label", pyarrow.large_string()),
        ("trip_update.vehicle.license_plate", pyarrow.large_string()),
        # trip_update.vehicle.consist: list<element: struct<label: string>>
        #   child 0, element: struct<label: string>
        #       child 0, label: string
        ("trip_update.vehicle.assignment_status", pyarrow.large_string()),
        ("trip_update.timestamp", pyarrow.timestamp("us")),
        ("trip_update.delay", pyarrow.int32()),
        ("feed_timestamp", pyarrow.timestamp("us")),
        ("trip_update.stop_time_update.stop_sequence", pyarrow.uint32()),
        ("trip_update.stop_time_update.stop_id", pyarrow.large_string()),
        ("trip_update.stop_time_update.arrival.delay", pyarrow.int32()),
        ("trip_update.stop_time_update.arrival.time", pyarrow.timestamp("us")),
        ("trip_update.stop_time_update.arrival.uncertainty", pyarrow.int32()),
        ("trip_update.stop_time_update.departure.delay", pyarrow.int32()),
        ("trip_update.stop_time_update.departure.time", pyarrow.timestamp("us")),
        ("trip_update.stop_time_update.departure.uncertainty", pyarrow.int32()),
        ("trip_update.stop_time_update.schedule_relationship", pyarrow.large_string()),
        ("trip_update.stop_time_update.boarding_status", pyarrow.large_string()),
        ("feed_timestamp_first_prediction", pyarrow.timestamp("us")),
        ("feed_timestamp_last_prediction", pyarrow.timestamp("us")),
    ]
)


def lrtp_prod(polars_df: pl.DataFrame) -> pl.DataFrame:
    """
    Function to apply final conversions to lamp data before outputting for tableau consumption
    """
    polars_df = polars_df.filter(
        pl.col("trip_update.stop_time_update.stop_id").is_in(LightRailFilter.terminal_stop_ids)
    )
    polars_df = apply_timezone_conversions(polars_df)
    return polars_df


def lrtp_devgreen(polars_df: pl.DataFrame) -> pl.DataFrame:
    """
    Function to apply final conversions to lamp data before outputting for tableau consumption
    This is intended for more complicated transformations than is feasible to perform in pyarrow

    Parameters
    ----------
    polars_df : Dataframe filtered down to light rail trip updates

    Returns
    -------
    pl.Dataframe : filtered down to departures at terminals
                   add feed_timestamp.first_prediction and feed_timestamp.last_prediction columns
                   to signify a validity duration of a prediction

    """

    # filter down to only terminals - original data
    polars_df = polars_df.filter(
        ~pl.col("trip_update.stop_time_update.departure.time").is_null()
        & pl.col("trip_update.stop_time_update.stop_id").is_in(LightRailFilter.terminal_stop_ids)
    )
    polars_df = append_prediction_valid_duration(polars_df)
    polars_df = apply_timezone_conversions(polars_df)
    return polars_df


def heavyrail(polars_df: pl.DataFrame) -> pl.DataFrame:
    """
    Function to apply final conversions to lamp data before outputting for tableau consumption
    """

    polars_df = polars_df.filter(
        ~pl.col("trip_update.stop_time_update.departure.time").is_null()
        & pl.col("trip_update.stop_time_update.stop_id").is_in(HeavyRailFilter.terminal_stop_ids)
    )
    polars_df = apply_timezone_conversions(polars_df)
    return polars_df


def apply_timezone_conversions(polars_df: pl.DataFrame) -> pl.DataFrame:
    """
    Function to apply timezone conversions to lamp data before outputting for tableau consumption
    """
    polars_df = polars_df.with_columns(
        pl.from_epoch(pl.col("trip_update.stop_time_update.departure.time"), time_unit="s")
        .dt.convert_time_zone(time_zone="US/Eastern")
        .dt.replace_time_zone(None),
        pl.from_epoch(pl.col("trip_update.stop_time_update.arrival.time"), time_unit="s")
        .dt.convert_time_zone(time_zone="US/Eastern")
        .dt.replace_time_zone(None),
        pl.from_epoch(pl.col("trip_update.timestamp"), time_unit="s")
        .dt.convert_time_zone(time_zone="US/Eastern")
        .dt.replace_time_zone(None),
        pl.from_epoch(pl.col("feed_timestamp"), time_unit="s")
        .dt.convert_time_zone(time_zone="US/Eastern")
        .dt.replace_time_zone(None),
        pl.from_epoch(pl.col("feed_timestamp_first_prediction"), time_unit="s")
        .dt.convert_time_zone(time_zone="US/Eastern")
        .dt.replace_time_zone(None),
        pl.from_epoch(pl.col("feed_timestamp_last_prediction"), time_unit="s")
        .dt.convert_time_zone(time_zone="US/Eastern")
        .dt.replace_time_zone(None),
    )
    return polars_df


def schema() -> pyarrow.schema:
    """
    Function returns schema of the processed gtfs-rt trip updates
    """
    return gtfs_rt_trip_updates_processed_schema


def append_prediction_valid_duration(polars_df: pl.DataFrame) -> pl.DataFrame:
    """
    Append feed_timestamp_first_prediction and feed_timestamp_last_prediction columns to the dataframe

    Predictions are valid only instantaneously from the upstream producer (RTR)
    This method attempts to derive a rough "validity period" by checking when the
    first prediction is made vs the last one grouped by trip_id, feed_timestamp, and predicted time

    The intent is to isolate timestamps that are sent by RTR that are meant for the
    same trip_id, but has not changed the prediction (departure.time) since the previous query.

    Parameters
    ----------
    polars_df : Dataframe filtered down to light rail trip updates

    Returns
    -------
    pl.Dataframe : dataframe with feed_timestamp_first_prediction and feed_timestamp_last_prediction columns
                   to signify a validity duration of a prediction
                   these columns are all null if enable_calculation argument is False
                   these columns are calculated and valid if enable_calculation is True
    """

    prediction_valid_duration = polars_df.group_by(
        ["trip_update.trip.trip_id", "trip_update.timestamp", "trip_update.stop_time_update.departure.time"]
    ).agg(
        pl.col("feed_timestamp").min().alias("feed_timestamp_first_prediction"),
        pl.col("feed_timestamp").max().alias("feed_timestamp_last_prediction"),
    )

    polars_df = polars_df.join(
        prediction_valid_duration,
        on=["trip_update.trip.trip_id", "trip_update.timestamp", "trip_update.stop_time_update.departure.time"],
        how="inner",
        coalesce=True,
    ).sort(["trip_update.trip.trip_id", "trip_update.stop_time_update.stop_sequence", "feed_timestamp"])

    return polars_df
