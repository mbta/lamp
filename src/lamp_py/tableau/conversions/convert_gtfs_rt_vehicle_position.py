import polars as pl
import pyarrow

gtfs_rt_vehicle_positions_processed_schema = pyarrow.schema(
    [
        ("id", pyarrow.large_string()),
        ("vehicle.trip.trip_id", pyarrow.large_string()),
        ("vehicle.trip.route_id", pyarrow.large_string()),
        ("vehicle.trip.direction_id", pyarrow.uint8()),
        ("vehicle.trip.start_time", pyarrow.large_string()),
        ("vehicle.trip.start_date", pyarrow.large_string()),
        ("vehicle.trip.schedule_relationship", pyarrow.large_string()),
        ("vehicle.trip.route_pattern_id", pyarrow.large_string()),
        ("vehicle.trip.tm_trip_id", pyarrow.large_string()),
        ("vehicle.trip.overload_id", pyarrow.int64()),
        ("vehicle.trip.overload_offset", pyarrow.int64()),
        ("vehicle.trip.revenue", pyarrow.bool_()),
        ("vehicle.trip.last_trip", pyarrow.bool_()),
        ("vehicle.vehicle.id", pyarrow.large_string()),
        ("vehicle.vehicle.label", pyarrow.large_string()),
        ("vehicle.vehicle.license_plate", pyarrow.large_string()),
        # vehicle.vehicle.consist: list<element: struct<label: string>>
        #   child 0, element: struct<label: string>
        #       child 0, label: string
        ("vehicle.vehicle.assignment_status", pyarrow.large_string()),
        ("vehicle.position.bearing", pyarrow.uint16()),
        ("vehicle.position.latitude", pyarrow.float64()),
        ("vehicle.position.longitude", pyarrow.float64()),
        ("vehicle.position.speed", pyarrow.float64()),
        ("vehicle.position.odometer", pyarrow.float64()),
        ("vehicle.current_stop_sequence", pyarrow.uint32()),
        ("vehicle.stop_id", pyarrow.large_string()),
        ("vehicle.current_status", pyarrow.large_string()),
        ("vehicle.timestamp", pyarrow.timestamp("us")),
        ("vehicle.congestion_level", pyarrow.large_string()),
        ("vehicle.occupancy_status", pyarrow.large_string()),
        ("vehicle.occupancy_percentage", pyarrow.uint32()),
        # vehicle.multi_carriage_details: list<element: struct<id: string, label: string, occupancy_status: string, occupancy_percentage: int32, carriage_sequence: uint32>>
        #   child 0, element: struct<id: string, label: string, occupancy_status: string, occupancy_percentage: int32, carriage_sequence: uint32>
        #       child 0, id: string
        #       child 1, label: string
        #       child 2, occupancy_status: string
        #       child 3, occupancy_percentage: int32
        #       child 4, carriage_sequence: uint32
        ("feed_timestamp", pyarrow.timestamp("us")),
    ]
)


def lrtp(polars_df: pl.DataFrame) -> pl.DataFrame:
    """
    Function to apply final conversions to lamp data before outputting for tableau consumption
    """

    def lrtp_restrict_vp_to_only_terminal_stop_ids(polars_df: pl.DataFrame) -> pl.DataFrame:
        """
        Function to apply lrtp filters conversions to lamp data before outputting for tableau consumption
        """
        stop_ids = list(map(str, [70110, 70162, 70236, 70274, 70502, 70510]))

        #    pylint: disable=singleton-comparison
        polars_df = polars_df.filter(~pl.col("vehicle.timestamp").is_null() & pl.col("vehicle.stop_id").is_in(stop_ids))
        return polars_df

    def temporary_lrtp_assign_new_trip_ids(polars_df: pl.DataFrame, threshold_sec: int = 60 * 15) -> pl.DataFrame:
        """
        Function to apply temporary trip ids to trips that have the same trip_id assigned for improbable stop durations at a single station

        THIS IS NOT PERMANENT - THIS SHOULD BE REMOVED WHEN THE TRIP_IDS ARE PROPERLY POPULATED UPSTREAM

        """
        polars_df = (
            polars_df.sort("feed_timestamp")
            .with_columns(
                pl.col("feed_timestamp").diff().alias("mdiff").over("vehicle.trip.trip_id", "vehicle.trip.start_date")
            )
            .with_columns(
                pl.when(pl.col("mdiff") < threshold_sec)
                .then(0)
                .otherwise(1)
                .fill_null(0)
                .cum_sum()
                .cast(pl.String)
                .alias("new_id")
                .over("vehicle.trip.trip_id", "vehicle.trip.start_date")
            )
            .with_columns(
                pl.when(pl.col("new_id").ne("1"))
                .then(
                    pl.concat_str(
                        [pl.col("vehicle.trip.trip_id"), pl.lit("_LAMP"), pl.col("new_id")], ignore_nulls=True
                    )
                )
                .alias("vehicle.trip.trip_id1")
            )
            .with_columns(pl.coalesce("vehicle.trip.trip_id1", "vehicle.trip.trip_id").alias("vehicle.trip.trip_id"))
            .drop("vehicle.trip.trip_id1")
        )
        return polars_df

    polars_df = lrtp_restrict_vp_to_only_terminal_stop_ids(polars_df)
    # after we have filtered to only terminal stop_ids, then check that the trip_id vs timestamps make sense, and
    # assign new trip IDs if it doesn't
    polars_df = temporary_lrtp_assign_new_trip_ids(polars_df)
    polars_df = apply_gtfs_rt_vehicle_positions_timezone_conversions(polars_df)

    return polars_df


def heavyrail(polars_df: pl.DataFrame) -> pl.DataFrame:
    """
    Function to apply final conversions to lamp data before outputting for tableau consumption
    """
    polars_df = apply_gtfs_rt_vehicle_positions_timezone_conversions(polars_df)
    polars_df = apply_gtfs_vehicle_positions_heavy(polars_df)

    return polars_df


def apply_gtfs_vehicle_positions_heavy(polars_df: pl.DataFrame) -> pl.DataFrame:
    """
    Function to apply lrtp filters conversions to lamp data before outputting for tableau consumption
    """
    stop_ids = list(map(str, [70003, 70034, 70040, 70057, 70063, 70092, 70104]))

    #    pylint: disable=singleton-comparison
    polars_df = polars_df.filter(~pl.col("vehicle.timestamp").is_null() & pl.col("vehicle.stop_id").is_in(stop_ids))
    return polars_df


def apply_gtfs_rt_vehicle_positions_timezone_conversions(polars_df: pl.DataFrame) -> pl.DataFrame:
    """
    Function to apply final conversions to lamp data before outputting for tableau consumption
    """
    polars_df = polars_df.with_columns(
        pl.from_epoch(pl.col("vehicle.timestamp"), time_unit="s")
        .dt.convert_time_zone(time_zone="US/Eastern")
        .dt.replace_time_zone(None),
        pl.from_epoch(pl.col("feed_timestamp"), time_unit="s")
        .dt.convert_time_zone(time_zone="US/Eastern")
        .dt.replace_time_zone(None),
    )
    return polars_df


def schema() -> pyarrow.schema:
    """
    Function returns schema of the processed gtfs-rt vehicle positions
    """
    return gtfs_rt_vehicle_positions_processed_schema
