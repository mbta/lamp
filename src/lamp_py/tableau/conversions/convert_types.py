from typing import Optional
import pyarrow


def convert_to_tableau_compatible_schema(
    input_schema: pyarrow.schema, overrides: Optional[dict] = None, exclude: Optional[list] = None
) -> pyarrow.schema:
    """
    generic converter for known types (dates, timezones, etc)
    overrides to correct the rest
    overrides_exclude -
    """
    # Create a map to store the modified fields that can't be auto-inferred
    auto_schema = []
    # Loop through the schema fields and apply changes
    for field in input_schema:
        print(field.type)

        if exclude is not None and field.name in exclude:
            continue
        if overrides is not None and field.name in overrides.keys():
            new_field = field.with_type(overrides[field.name])
            auto_schema.append(new_field)
            continue
        if isinstance(field.type, pyarrow.TimestampType):
            if field.type.tz is not None:
                # strip tz
                new_field = field.with_type(pyarrow.timestamp(unit=field.type.unit))
                auto_schema.append(new_field)
                continue

        # default - no changes, add the field again
        auto_schema.append(field)

    return pyarrow.schema(auto_schema)
