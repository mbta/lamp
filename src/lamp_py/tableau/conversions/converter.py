import os
from abc import ABC
from abc import abstractmethod
from itertools import chain
from typing import Dict
from typing import Optional

import pyarrow
from pyarrow import fs
import pyarrow.parquet as pq

from lamp_py.postgres.postgres_utils import DatabaseManager
from lamp_py.runtime_utils.process_logger import ProcessLogger
from lamp_py.aws.s3 import (
    download_file,
    upload_file,
    object_metadata,
)

class Converter(ABC):  # pylint: disable=R0902
    """
    Abstract Base Class for Parquet jobs
    """

    def __init__(
        self,
        hyper_file_name: str,
        remote_parquet_path: str,
        lamp_version: str,
        project_name: str = os.getenv("TABLEAU_PROJECT", ""),
    ) -> None:
        environment = os.getenv("ECS_TASK_GROUP", "-").split("-")[-1]
        if environment != "prod":
            hyper_file_name = f"{hyper_file_name.replace('.hyper','')}_{environment}.hyper"
        self.hyper_file_name = hyper_file_name

        self.hyper_table_name = hyper_file_name.replace(".hyper", "")
        self.remote_parquet_path = remote_parquet_path
        self.lamp_version = lamp_version
        self.project_name = project_name
        self.local_parquet_path = "/tmp/local.parquet"
        self.local_hyper_path = f"/tmp/{hyper_file_name}"

        self.remote_fs = fs.LocalFileSystem()
        if remote_parquet_path.startswith("s3://"):
            self.remote_fs = fs.S3FileSystem()
            self.remote_parquet_path = self.remote_parquet_path.replace("s3://", "")

    @property
    @abstractmethod
    def parquet_schema(self) -> pyarrow.schema:
        """
        Schema for Job parquet file
        """

    @abstractmethod
    def create_parquet(self, db_manager: Optional[DatabaseManager]) -> None:
        """
        Business logic to create new Job parquet file
        """

    @abstractmethod
    def update_parquet(self, db_manager: Optional[DatabaseManager]) -> bool:
        """
        Business logic to update existing Job parquet file

        :return: True if local parquet file was updated, else False
        """

    def max_stats_of_parquet(self) -> Dict[str, str]:
        """
        Create dictionary of maximum value for each column of locally saved
        parquet file

        :return Dict[column_name: max_column_value]
        """
        # get row_groups from parquet metadata (list of dicts)
        row_groups = pq.read_metadata(self.local_parquet_path).to_dict()["row_groups"]

        # explode columns element from all row  groups into flat list
        parquet_column_stats = list(chain.from_iterable([row_group["columns"] for row_group in row_groups]))

        return {
            col["path_in_schema"]: col["statistics"].get("max") for col in parquet_column_stats if col["statistics"]
        }

    def remote_version_match(self) -> bool:
        """
        Compare "lamp_version" of remote parquet file to expected.

        :return True if remote and expected version match, else False
        """
        lamp_version = object_metadata(self.remote_parquet_path).get("lamp_version", "")

        return lamp_version == self.lamp_version

    def run_parquet(self, db_manager: Optional[DatabaseManager]) -> None:
        """
        Remote parquet Create / Update runner

        If Remote parquet file does not exist, or Job parquet schema does not
        match remote parquet schema, create new remote parquet file.

        Otherwise, update existing remote parquet file
        """
        process_log = ProcessLogger(
            "hyper_job_run_parquet",
            remote_path=self.remote_parquet_path,
        )
        process_log.log_start()

        try:
            remote_schema_match = False
            remote_version_match = False
            file_info = self.remote_fs.get_file_info(self.remote_parquet_path)

            if file_info.type == fs.FileType.File:
                # get remote parquet schema and compare to expected local schema
                remote_schema = pq.read_schema(
                    self.remote_parquet_path,
                    filesystem=self.remote_fs,
                )
                remote_schema_match = self.parquet_schema.equals(remote_schema)
                remote_version_match = self.remote_version_match()

            if remote_schema_match is False or remote_version_match is False:
                # create new parquet if no remote parquet found or
                # remote schema does not match expected local schema
                run_action = "create"
                upload_parquet = True
                self.create_parquet(db_manager)
            else:
                run_action = "update"
                upload_parquet = self.update_parquet(db_manager)

            parquet_file_size_mb = 0.0
            if os.path.exists(self.local_parquet_path):
                parquet_file_size_mb = os.path.getsize(self.local_parquet_path) / (1024 * 1024)

            process_log.add_metadata(
                remote_schema_match=remote_schema_match,
                run_action=run_action,
                upload_parquet=upload_parquet,
                parquet_file_size_mb=f"{parquet_file_size_mb:.2f}",
            )

            if upload_parquet:
                upload_file(
                    file_name=self.local_parquet_path,
                    object_path=self.remote_parquet_path,
                    extra_args={"Metadata": {"lamp_version": self.lamp_version}},
                )

            if os.path.exists(self.local_parquet_path):
                os.remove(self.local_parquet_path)

            process_log.log_complete()

        except Exception as exception:
            process_log.log_failure(exception=exception)
