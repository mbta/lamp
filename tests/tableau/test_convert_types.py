import pyarrow
import pytest
from lamp_py.tableau.conversions.convert_types import convert_to_tableau_compatible_schema


# simple case
@pytest.fixture
def test_schema1() -> pyarrow.schema:
    return pyarrow.schema(
        [
            ("service_date", pyarrow.string()),
            ("route_id", pyarrow.large_string()),
            ("exact_plan_trip_match", pyarrow.bool_()),
        ]
    )


# simple case
@pytest.fixture
def test_schema2() -> pyarrow.schema:
    return pyarrow.schema(
        [
            ("service_date", pyarrow.date32()),
            ("route_id", pyarrow.large_string()),
            ("exact_plan_trip_match", pyarrow.bool_()),
            ("start_dt", pyarrow.timestamp(unit="us", tz="UTC")),
        ]
    )


# all case
@pytest.fixture
def test_schema3() -> pyarrow.schema:
    return pyarrow.schema(
        [
            ("service_date", pyarrow.string()),
            ("route_id", pyarrow.large_string()),
            ("exact_plan_trip_match", pyarrow.bool_()),
            ("start_dt", pyarrow.timestamp(unit="us", tz="UTC")),
        ]
    )


def test_tableau_auto_schema_happy_case(test_schema1: pyarrow.schema) -> None:
    assert test_schema1 == convert_to_tableau_compatible_schema(test_schema1)


def test_tableau_auto_schema_override(test_schema1: pyarrow.schema) -> None:
    assert_test_schema1 = pyarrow.schema(
        [
            ("service_date", pyarrow.date32()),  # change to date type
            ("route_id", pyarrow.large_string()),
            ("exact_plan_trip_match", pyarrow.bool_()),
        ]
    )
    overrides = {"service_date": pyarrow.date32()}
    assert assert_test_schema1 == convert_to_tableau_compatible_schema(test_schema1, overrides=overrides)


def test_tableau_auto_schema_exclude(test_schema1: pyarrow.schema) -> None:
    assert_test_schema1 = pyarrow.schema(
        [
            ("exact_plan_trip_match", pyarrow.bool_()),
        ]
    )
    excludes = ["service_date", "route_id"]
    assert assert_test_schema1 == convert_to_tableau_compatible_schema(test_schema1, exclude=excludes)


def test_tableau_auto_schema_tz(test_schema2: pyarrow.schema) -> None:
    assert_test_schema2 = pyarrow.schema(
        [
            ("service_date", pyarrow.date32()),  # change to date type
            ("route_id", pyarrow.large_string()),
            ("exact_plan_trip_match", pyarrow.bool_()),
            ("start_dt", pyarrow.timestamp(unit="us", tz=None)),
        ]
    )
    assert assert_test_schema2 == convert_to_tableau_compatible_schema(test_schema2)


def test_tableau_auto_schema_all_case(test_schema3: pyarrow.schema) -> None:
    breakpoint()
    assert_test_schema3 = pyarrow.schema(
        [
            ("service_date", pyarrow.date32()),  # change to date type
            ("exact_plan_trip_match", pyarrow.bool_()),
            ("start_dt", pyarrow.timestamp(unit="us", tz=None)),
        ]
    )
    overrides = {"service_date": pyarrow.date32()}
    excludes = ["route_id"]
    assert assert_test_schema3 == convert_to_tableau_compatible_schema(
        test_schema3, overrides=overrides, exclude=excludes
    )
