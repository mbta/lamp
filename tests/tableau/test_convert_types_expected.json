[
    {
        "data": "{\"name\":\"Alice\",\"age\":30},{\"name\":\"Bob\",\"age\":25}",
        "data2": "{\"name\":\"Darcy\",\"age\":30},{\"name\":\"Effie\",\"age\":25}"
    },
    {
        "data": "{\"name\":\"Charlie\",\"age\":40}",
        "data2": "{\"name\":\"Gerard\",\"age\":40},{\"name\":\"Henry\",\"age\":30},{\"name\":\"Iggy\",\"age\":25}"
    },
    {
        "data": "",
        "data2": "{\"name\":\"Juliet\",\"age\":40}"
    }
]